README.md

```diff
@@ -2,43 +2,193 @@
+          #          hash from the error message
+        }).rust;
+
+        naersk' = pkgs.callPackage naersk {
+          cargo = toolchain;
+          rustc = toolchain;
+        };
+        
+      in rec {
+        # For `nix build` & `nix run`:
+        defaultPackage = naersk'.buildPackage {
+          src = ./.;
+        };
+
+        # For `nix develop` (optional, can be skipped):
+        devShell = pkgs.mkShell {
+          nativeBuildInputs = [ toolchain ];
+        };
+      }
+    );
+}
+```
+
+### Using Niv
+
+``` shell
+$ niv init
 $ niv add nix-community/naersk
 ```
 
-And then
+... and then create `default.nix` with:
 
 ``` nix
 let
-    pkgs = import <nixpkgs> {};
-    sources = import ./nix/sources.nix;
-    naersk = pkgs.callPackage sources.naersk {};
-in naersk.buildPackage ./path/to/rust
+  pkgs = import <nixpkgs> {};
+  sources = import ./nix/sources.nix;
+  naersk = pkgs.callPackage sources.naersk {};
+  
+in
+  naersk.buildPackage ./.
 ```
 
-_NOTE_: `./path/to/rust/` should contain a `Cargo.lock`.
+(this assumes `default.nix` is created next to `Cargo.toml` & `Cargo.lock` - if
+that's not the case, adjust `./.` in `naersk.buildPackage`.)
 
-## Configuration
+Note that Naersk by default ignores the `rust-toolchain` file, using whatever's
+Rust compiler version is present in `nixpkgs`.
 
-The `buildPackage` function also accepts an attribute set. The attributes are
-described below. Any attribute that is _not_ listed below will be forwarded _as
-is_ to `stdenv.mkDerivation`. When the argument passed in _not_ an attribute
-set, e.g.
+If you've got custom `rust-toolchain` file, you can make Naersk use it this way:
+
+``` shell
+$ niv add mozilla/nixpkgs-mozilla
+```
+
+... and then:
 
 ``` nix
-naersk.buildPackage theArg
+let
+  sources = import ./nix/sources.nix;
+  nixpkgs-mozilla = import sources.nixpkgs-mozilla;
+  
+  pkgs = import sources.nixpkgs {
+    overlays = [
+      nixpkgs-mozilla
+    ];
+  };
+  
+  toolchain = (pkgs.rustChannelOf {
+    rustToolchain = ./rust-toolchain;
+    sha256 = "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=";
+    #        ^ After you run `nix-build`, replace this with the actual
+    #          hash from the error message
+  }).rust;
+  
+  naersk = pkgs.callPackage sources.naersk {
+    cargo = toolchain;
+    rustc = toolchain;
+  };
+  
+in
+  naersk.buildPackage ./.
 ```
 
-it is converted to an attribute set equivalent to `{ root = theArg; }`.
+## Usage
+
+Naersk provides a function called `buildPackage` that takes an attribute set
+describing your application's directory, its dependencies etc.; in general, the
+usage is:
+
+``` nix
+naersk.buildPackage {
+  # Assuming there's `Cargo.toml` right in this directory:
+  src = ./.; 
+  
+  someOption = "yass";
+  someOtherOption = false;
+  CARGO_ENVIRONMENTAL_VARIABLE = "test";
+}
+```
+
+Some of the options (described below) are used by Naersk to affect the building
+process, rest is passed-through into `mkDerivation`.
```

Comment:
is it correct that the rest is passed to `mkDerivation`? I remember it didn't use to do that but maybe that's changed

Reply:
I think so (see: `drv = stdenv.mkDerivation (drvAttrs // userAttrs);`) - otherwise things such as environmental variables and custom build hooks (except `postInstall`, which we support natively now) wouldn't really work, right? 👀